src/main.rs

```diff
@@ -32,30 +40,40 @@ async fn main() -> anyhow::Result<()> {
     Ok(())
 }
 
-async fn server_process(transport: &mut Framed<TcpStream, NetstringCodec>, m: &Bytes) -> Result<()> {
+fn server_process(transport: Pin<&mut SplitSink<Framed<TcpStream, NetstringCodec>, Bytes>>, m: &Bytes, cx: &mut Context<'_>) -> Result<()> {
     let mut buf = BytesMut::with_capacity(m.len() + 20);
     buf.extend_from_slice(&m);
     buf.extend_from_slice(b" <- got this, i am echoing!");
-    transport.send(buf.freeze()).await?;
+    transport.start_send(buf.freeze());
     Ok(())
 }
 
+struct Receiver {
+    // frame: Framed<TcpStream, NetstringCodec>,
+    sink: SplitSink<Framed<TcpStream, NetstringCodec>, Bytes>,
+    stream: SplitStream<Framed<TcpStream, NetstringCodec>>,
+}
+
+impl Future for Receiver {
+    type Output = Result<BytesMut, std::io::Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        while let Poll::Ready(res) = Pin::new(&mut self.stream).poll_next(cx) {
```

Comment:
i am worried that we are only hiding the polling here.
here are some readings:

### ideal conditions:
-> https://docs.rs/futures-core/0.3.17/futures_core/stream/trait.Stream.html
here is, under an ideal conditions, a stream when we have a proper stream - it sends 
Pending if the stream is active and there is no item available, Ready(Some(x)) when is active and item is available and Ready(None) when stream has been terminated

### what we have:
-> https://docs.rs/tokio/0.2.9/tokio/stream/trait.StreamExt.html#method.next
now our problem is that we are using the stream as an iterator, which stream technically is, and we are calling .next() on it, according to the provided documentation it is behaving as it should, Ready(None) means stream is exhausted, but not really broken. IMO it means that either tokio or rust allows transforming stream into iterator without saying "ey, mate, hold up". This can possibly result in perfectly sound code that is compiling, working, and being statically analyzable, but is indeed broken inside (like me :3). 

However this is what tokio provides as an example for us, so i am really confused and do not really know how to make it work. Technically this should be what we should call:
-> https://docs.rs/futures-sink/0.3.17/futures_sink/trait.Sink.html#tymethod.poll_ready


### in free time
exercise to be had: instead of fragmented matching ( `let Some(x) =...` ) try to replace it with full proper match and emit logs on each misfire, i.e. catch "None" loudly. This will tell us if its working as intended or its doing what i did at the beginning and its just spinning like stupid silently.

Reply:
`in free time` I've pushed new version which `break`s when Ready(None) is received and it waits until new connection. There still may be something not working if we keep connection alive though